docs/latest/design/ratelimit.md

```diff
@@ -0,0 +1,272 @@
+* A `match` holds true, when all conditions under the `match` hold true.
+
+Here's an example highlighting this -
+
+```
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: RateLimitFilter
+metadata:
+  name: ratelimit-all-safeguard-app 
+spec:
+  type: Global
+  global:
+    rules:
+    - matches:
+      - limit:
+          requests: 100
+          unit: Second
+---
+
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: RateLimitFilter
+metadata:
+  name: ratelimit-per-user
+spec:
+  type: Global
+  global:
+    rules:
+    - matches:
+      - Type: Distinct
+        header:
+          name: x-user-id
+        limit:
+          requests: 1000
+          unit: Hour
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: example
+spec:
+  parentRefs:
+    - name: eg
+  hostnames:
+    - www.example.com
+  rules:
+    - matches:
+        - path:
+            type: PathPrefix
+            value: /foo
+      filters:
+        - type: ExtensionRef
+          extensionRef:
+            group: gateway.envoyproxy.io
+            kind: RateLimitFilter
+            name: ratelimit-per-user
+	- type: ExtensionRef
+          extensionRef:
+            group: gateway.envoyproxy.io
+            kind: RateLimitFilter
+            name: ratelimit-all-safeguard-app    
+      backendRefs:
+        - name: backend
+          port: 3000
+```
+
+* The user has created two `RateLimitFilter`s  and has attached it to a `HTTPRoute` - one(`ratelimit-all-safeguard-app`) to
+ensure that the backend does not get overwhelmed with requests, any excess requests are rate limited irrespective of
+the attributes within the traffic flow, and another(`ratelimit-per-user`) to rate limit each distinct user client
+who can be differentiated using the `x-user-id` header, to ensure that each client does not make exessive requests to the backend.
+* If user `baz` (identified with the header and value of `x-user-id: baz`) sends 90 requests within the first second, and
+user `bar` sends 11 more requests during that same interval of 1 second, and user `bar` sends the 101th request within that second,
+the rule defined in `ratelimit-all-safeguard-app` gets activated and Envoy Gateway will ratelimit the request sent by `bar` (and any other
+request sent within that 1 second). After 1 second, the rate limit counter associated with the `ratelimit-all-safeguard-app` rule
+is reset and again evaluated.
+* If user `bar` also ends up sending 90 more requests within the hour, summing up `bar`'s total request count to 101, the rate limit rule
+defined within `ratelimit-per-user` will get activated, and `bar`'s requests will be rate limited again until the hour interval ends.
+* Within the same above hour, if `baz` sends 991 more requests, summing up `baz`'s total request count to 1001, the rate limit rule defined
+within `ratelimit-per-user` will get activated for `baz`, and `baz`'s requests will also be rate limited until the hour interval ends. 
+
```

Comment:
Thanks for this! It may seem trivial but it's a huge help.

I think that there're a couple of other things that must be discussed here:

1. As described, you're setting up ratelimits as a strict any-of operation: if _any_ ratelimit fires, the request is ratelimited. If that's what you mean, I think it should be explicitly called out, so that it's obvious to the reader.

2. How about precedence and ordering? Is the first match the rule that fires, or is the best match? For a concrete example here, how would you set up ratelimiting to allow unauthenticated users 1RPS, but most authenticated users 10RPS, but authenticated users Alice and Bob 100RPS?

Reply:
thanks for raising these points, 
1. you're right, ratelimits are setup like `any-of` op, will ensure this is explicitly stated in the API doc.
2. Since its an `any-of` op, each limit would be applied in a mutually exclusive way. To ensure Alice and Bob's traffic dont get limited at 10RPS, but 100 RPS instead, we will either need to use a separate attribute to select authenticated users, or we will need to use some *NOT* match logic to ensure Bob and Alice dont get limited at 10RPS, this isn't of the API today, but can be added in the future (envoy has it internally, is called [expect_match](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-ratelimit-action-headervaluematch))

Reply:
there is no notion of precedence in this API, ive pushed https://github.com/envoyproxy/gateway/pull/706/commits/ad5627964309cb0d6aa86d408f2b683fb359168d to highlight it